### PR TITLE
Cja 34 add routing

### PR DESCRIPTION
**Pull request by frankiebarrios**
## What It Does
This pull request is set to complete Add Routing.

This work is taking place to complete Add Routing.

**Implementation:**
- Routing setup using the Banno Router.
- Separates the app's functionalities into views.
- Adds routing between views.
- Uses routes as a source of data to populate your different views.

## How To Test
- Install dependencies using instruction in README.
- Start server using yarn start or build using yarn build.

## Notes/Caveats

None
## Checklist

Under penalty of public shaming, I avow that I:
**Many items in list N/A** 
- [X] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [X] Verified that there are no compiler or linter errors or warnings
- [X] Verified the build in production(optimized) mode
- [X] Updated the docs, if necessary
- [X] Included the appropriate labels
- [X] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
**Many items in list N/A** 
- [X] Chrome
- [X] Firefox
- [X] Safari
- [X] Edge
- [ ] Internet Explorer

## Dependencies
1. Install Node and npm
2. Install Yarn
3. Install Ux-Lint

### CODE REVIEW

**Review by kddrake — COMMENTED** [review]
what's all this `.idea/*` stuff?

File: src/components/create-user.html [hunk 1]
```diff
@@ -109,12 +109,24 @@ <h5>Create</h5>
   <script type="module">
     import { Element as PolymerElement } from '@banno/polymer/polymer-element.js';
 
-    class CreateUser extends PolymerElement {
+    // Routing Imports
```
Comment (kddrake) [inline src/components/create-user.html]: your imports are named well... I don't think you need the comment.

File: src/components/create-user.html [hunk 2]
```diff
@@ -109,12 +109,24 @@ <h5>Create</h5>
...
+
+    class CreateUser extends bindPropertiesFromParentRouteMixin(RouteMixin(PolymerElement)) {
       static get is() { return 'create-user'; }
 
       static get properties() {
         return {
-          storage: Object,
+          storage: {
+            type: Object,
+            observer: 'check'
```
Comment (kddrake) [inline src/components/create-user.html]: probably nice when deving, but I don't think you'll actually want to merge this.

also, you could do this all in one line:
```suggestion
            observer: (newVal, oldVal) => console.log(newVal)
```

File: src/components/create-user.html [hunk 3]
```diff
@@ -174,21 +186,44 @@ <h5>Create</h5>
...
+        console.log('Create User Parent Node on Enter: ', currentElement.parentNode);
+        next();
+      }
+
+      routeExit(currentNode, nextNode, routeId, context, next) {
+        super.routeExit(currentNode, nextNode, routeId, context, next)
+        const currentElement = currentNode.getValue().element;
+        console.log('Create User Parent Node on Exit: ', currentElement.parentNode);
+        if (currentElement.parentNode) {
+          currentElement.parentNode.removeChild((currentElement));
```
Comment (kddrake) [inline src/components/create-user.html]: why the double wrap of `currentElement`?

File: src/components/intern-app.html [hunk 4]
```diff
@@ -56,7 +100,15 @@
 
     const storage = new FirebaseAPI();
 
-    class InternApp extends PolymerElement {
+    // Routing Imports
```
Comment (kddrake) [inline src/components/intern-app.html]: see other comment about this same comment

File: src/components/user-list.html [hunk 5]
```diff
@@ -8,23 +8,57 @@
...
+        <user-profile user="[[user]]" storage="[[storage]]"></user-profile>
       </template>
     </div>
   </template>
 
   <script type="module">
     import { Element as PolymerElement } from '@banno/polymer/polymer-element.js';
 
-    class UserList extends PolymerElement {
+    // Routing Imports
```
Comment (kddrake) [inline src/components/user-list.html]: see other comment about this comment

File: src/components/user-list.html [hunk 6]
```diff
@@ -8,23 +8,57 @@
...
+    class UserList extends bindPropertiesFromParentRouteMixin(RouteMixin(PolymerElement)) {
       static get is() { return 'user-list'; }
 
       static get properties() {
         return {
           users: Object,
-          storage: Object
+          storage: {
+            type: Object,
+            observer: 'check'
```
Comment (kddrake) [inline src/components/user-list.html]: see other comment about this observer

File: src/components/user-profile.html [hunk 7]
```diff
@@ -105,15 +105,27 @@ <h3 class="div-center" style="margin: initial;padding-bottom: inherit;">User Pro
...
+    class UserProfile extends bindPropertiesFromParentRouteMixin(RouteMixin(PolymerElement)) {
       static get is() { return 'user-profile'; }
 
       static get properties() {
         return {
-          storage: Object,
-          user: { 
+          storage: {
+            type: Object,
+            observer: 'check'
```
Comment (kddrake) [inline src/components/user-profile.html]: see other comment about this observer

**Comment by frankiebarrios** [comment]
> what's all this `.idea/*` stuff?

@kddrake it's an IntelliJ’s settings folder. Tried a new IDE from JetBrains and had to install some plugins/ extensions and it added that to my project. Deleting it though, not needed.

File: src/components/user-profile.html [hunk 8]
```diff
@@ -136,9 +171,9 @@ <h3 class="div-center" style="margin: initial;padding-bottom: inherit;">User Pro
 
       expandCollapseCard() {
         this.profileExpanded ? this.profileExpanded = false :
```
Comment (kddrake) [inline src/components/user-profile.html]: I think this is what you're going for, and much simpler:
```suggestion
        this.profileExpanded = !this.profileExpanded;
```

File: src/components/user-profile.html [hunk 9]
```diff
@@ -136,9 +171,9 @@ <h3 class="div-center" style="margin: initial;padding-bottom: inherit;">User Pro
...
-         this.profileExpanded = true;
+          this.profileExpanded = true;
         this.buttonText === 'Expand' ? this.buttonText = 'Collapse' :
```
Comment (kddrake) [inline src/components/user-profile.html]: Would do the same here, as well:
```suggestion
        this.buttonText = !this.buttonText;
```

File: src/routing/routes.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  const node = new RouteTreeNode(new RouteData(...nodeData.args));
+  node.title = nodeData.title;
+  node.inheritedProperties = nodeData.inheritedProperties;
+  node.navItems = nodeData.navItems;
+  if (nodeData.children) {
+    nodeData.children.forEach(nodeData => {
+      node.addChild(buildRouteTree(nodeData));
+    });
+  }
+  console.log('Tree: ', node);
```
Comment (kddrake) [inline src/routing/routes.js]: does this _need_ to be merged?

**Review by kddrake — CHANGES_REQUESTED** [review]
Things, code wise, are generally looking good. I will say that your `routeEnter` functions a bit more involved than most of our `routeEnter` functions in the platform. That said, if it works, it works. I'm not knowledgeable enough on the router to say whether or not there's a better way.

File: src/components/create-user.html (hunk 3 above)
Comment (frankiebarrios) [reply src/components/create-user.html]: Had something else in there at one time I believe? Removed and all is well still.

**Review by kddrake — APPROVED** [review]
I would have had some other changes I would have liked to have been made, but seeing as I want a focus on LitElement and you've proven you can get the router in your app and get it to work properly, I'ma approve and merge 👍

**Pull request merged**